plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs

```diff
@@ -62,39 +70,43 @@ nameNameSemanticFromHie hieKind rm ns = do
 -- | get only visible names from HieAST
 -- we care only the leaf node of the AST
 -- and filter out the derived and evidence names
-hieAstSpanNames :: VirtualFile -> HieAST a -> M.Map Range NameSet
-hieAstSpanNames vf ast =
+hieAstSpanIdentifiers :: VirtualFile -> HieAST a -> M.Map Range (Set Identifier)
+hieAstSpanIdentifiers vf ast =
   if null (nodeChildren ast)
     then getIds ast
-    else M.unionsWith unionNameSet $ map (hieAstSpanNames vf) (nodeChildren ast)
+    else M.unionsWith S.union $ map (hieAstSpanIdentifiers vf) (nodeChildren ast)
   where
     getIds ast' = fromMaybe mempty $ do
       range <- codePointRangeToRange vf $ realSrcSpanToCodePointRange $ nodeSpan ast'
-      return $ M.singleton range (getNodeIds' ast')
+      return $
+        M.fromListWith
+          (<>)
+          [S.singleton <$> ri | idt <- S.toList (getNodeIds' ast'), ri <- splitModuleNameAndOccName vf range idt]
     getNodeIds' =
       Map.foldl' combineNodeIds mempty
         . Map.filterWithKey (\k _ -> k == SourceInfo)
         . getSourcedNodeInfo
         . sourcedNodeInfo
-    combineNodeIds :: NameSet -> NodeInfo a -> NameSet
-    ad `combineNodeIds` (NodeInfo _ _ bd) = ad `unionNameSet` xs
+    combineNodeIds :: Set Identifier -> NodeInfo a -> Set Identifier
+    ad `combineNodeIds` (NodeInfo _ _ bd) = ad `S.union` xs
       where
-        xs = mkNameSet $ rights $ M.keys $ M.filterWithKey inclusion bd
+        xs = S.fromList $ M.keys $ M.filterWithKey inclusion bd
         inclusion :: Identifier -> IdentifierDetails a -> Bool
         inclusion a b = not $ exclusion a b
         exclusion :: Identifier -> IdentifierDetails a -> Bool
         exclusion idt IdentifierDetails {identInfo = infos} = case idt of
-          Left _  -> True
+          Left _  -> False
           Right _ -> any isEvidenceContext (S.toList infos)
```

Comment:
These cases could do with a comment explaining why we're excluding or not

Reply:
done